Basic support for running Emscripten code as AudioWorklets

tklajnscek opened the pull request:
Basically, this PR lets you run AudioWorkletNodes that can nicely call into Emscripten-generated code. 
It sets up a pthread environment with shared memory in the AudioWorkletGlobalScope that's shared by all the audio worklets.

### Summary of major changes:
- Implemented a futex that works in the worklets so that the mutex etc works nicely. It's the same busy wait as in the main thread, but without all the extra complexity (no proxying etc)
- Added slightly customized pthread bootstrapping for audio worklets. This works via a dummy AudioWorkletNode to get the message port for worker communication.
- All the script loading is done on the main thread in case of AudioWorklets
- Changed 'worker.js' to make it work in an AudioWorkletGlobalScope. This is mainly to stop it trying to import scripts from the worklet context (not allowed) and to replace various `this.XXX` since `this` doesn't exist in AudioWorkletGlobalScope

### Current limitations:
**Doesn't work with MINIMAL_RUNTIME** as I haven't figured out how to export functions so that they can be called from the registered `AudioWorkletProcessor`.

### Examples
There's an example in `tests/audioworklet` that also tests the pthread environment setup and the futex implementation.

We are using (a much less clean) version of this in production and it works great!
Any comments highly appreciated! :)

tklajnscek commented on emcc.py:
```diff
@@ -2856,18 +2856,28 @@ def modularize():
 
-    # Export using a UMD style export, or ES6 exports if selected
+    # Add ; if src doesn't contain it so we generate valid JS (otherwise acorn fails)
+    if src.rstrip()[-1] != ';':
```
This is pretty ugly, but there was a case where it would generate:
```
var Module = { all the code here }
if (typeof AudioWorkletGlobalScope === 'function')
...
```
which would fail as it passed through acorn because of that missing ;

I guess this needs fixing where it's generated... I'll check this again tomorrow.

tklajnscek commented on src/library_pthread.js:
```diff
@@ -388,7 +397,7 @@ var LibraryPThread = {
 
       worker.onerror = function(e) {
-        err('pthread sent an error! ' + e.filename + ':' + e.lineno + ': ' + e.message);
+        err('pthread sent an error! ' + e.stack ? e.stack : (e.filename + ':' + e.lineno + ': ' + e.message));
```
This makes the error output from unit tests/node much, much nicer (i.e. actually tells you what failed)

kripken commented:
Interesting @tklajnscek , but I'm a little lost here... can you please explain why this is needed? That is, is this because AudioWorklets are very different than normal environments (if so, what are the specific issues this fixes) or is this adding a new type of functionality entirely (using an AudioWorklet in some special way)?

Or to ask the question in another way, what does this make possible that wasn't before?

tklajnscek commented:
> Interesting @tklajnscek , but I'm a little lost here... can you please explain why this is needed? That is, is this because AudioWorklets are very different than normal environments (if so, what are the specific issues this fixes) or is this adding a new type of functionality entirely (using an AudioWorklet in some special way)?
> 
> Or to ask the question in another way, what does this make possible that wasn't before?

Sure, let me try to explain. Forgive me if some if this is stating the obvious 😅 

Let me run through the main pain points first vs regular Workers:
- The global scope is different and there are some things that don't work in AudioWorkletGlobalScope. This is mostly to make it difficult to stall the worklets as they are a real-time thing that needs to spit out the audio data ASAP. 
- So there's no `fetch`, `importScripts` etc and the only way to load scripts is by calling `addModule` on the main thread (this will also respect any **static** ES6 import statements in the loaded script). 
- There's also no `Atomics.wait` for fear of making it too easy to block 
- All AudioWorklets run serially in the browser's (single) audio thread and share the same global context (the AudioWorkletGlobalScope).
- Since the global scope is not shared their interface to the other threads is not on the global scope, like it is for Workers (postMessage/onmessage). The way AudioWorklets do it is they register classes derived from `AudioWorkletProcessor` when the script is first evaluated. Then instances of these classes are created by calling `new AudioWorkletNode(audioContext, 'name-of-processor')` on the main thread.

This PR addresses this by:
- Changing `worker.js` to work when evaluated in the AudioWorkletGlobalScope in addition to regular Workers
- Changes the loading/instantiation/bootstrapping code to play nicely with the fact that scripts are loaded via `addModule` avoiding calls to `importScripts` 
- If run in AudioWorkletGlobalScope it reroutes the `library_pthread` message passing through a dummy AudioWorkletProcessor which makes it work with the same code as regular Workers.
- Implements a special-case futex that can be used in AudioWorkletGlobalScope and uses a simple spin wait (like the main thread implementation) so that various synchronization primitives will work correctly.

All of the above makes it possible to run a shared memory instance of the program in AudioWorkletGlobalScope with the pthread environment initialized properly, just like a regular Worker. 

Which in turn makes it possible to execute all the program code in there which makes it super easy to run WASM code (=fast) in audio worklets and lets you port a bunch of existing C(++) audio stuff over much more easily.

The developer only needs to register their AudioWorkletProcessor and call their exported functions from it and it JustWorks(TM).

Without this you can't really run any Emscripten compiled code from an AudioWorkletProcessor unless you do **all the steps manually anyways** like transferring over the wasmInstance, getting the .js to evaluate, initializing the pthread environment manually to get the stack set up etc. So you end up running custom JS (or potentially, custom WASM+JS) in there with a bunch of glue code to get it to play along with your main program that does a bunch of processing already.

We currently use this to run the final output mixer as an AudioWorklet and it works great and it's the same code as any other platform (with the addition of the AudioWorkletProcessor class that's calling into it). In addition to this we run audio decoders as regular Workers.

I can also see this being very useful to anyone doing threaded/DPC audio output because it most likely maps 1:1 with what they have in there now.

I hope that answers the question @kripken. Let me know if anything is unclear or if I missed something obvious somewhere :)

kripken commented:
I see thanks @tklajnscek ! This does sound like a useful thing to support, and I think I understand now why it takes so much work here.

Some initial comments on the PR:

1. It looks like this duplicates a bunch of main thread logic, like having a special futex, busy-waiting when necessary, etc. It would be good to find a way to share that code as much as possible (like in wake, having a helper function that handles a special mutex, and it is called once for the main thread futex and once for the worklet).
2. I suspect we'd want to not include any worklet code when not building for that mode, so it should be ifdef'd out where feasible.
3. Adding a new valid environment value requires updating `settings.js` to mention it under `ENVIRONMENT`.

@juj , have you thought about worklet support?

tklajnscek commented:
@kripken I believe the latest changes should address all your points and it's rebased on latest master.
- I removed the code duplication in the futex implementation.
- If the ENVIRONMENT doesn't contain 'worklet' all the worklet support is stripped out (via ENVIRONMENT_MAY_BE_WORKLET)
- settings.js ENVIRONMENT entry has been updated

@sbc100 if you have any thoughts on this they'd be much appreciated :)

juj commented:
First off, thanks for working on this kind of project, this is a very interesting development, one which I have had a TODO for close to three(!) years now to experiment with..

> Which in turn makes it possible to execute all the program code in there which makes it super easy to run WASM code (=fast) in audio worklets and lets you port a bunch of existing C(++) audio stuff over much more easily.
> 
> The developer only needs to register their AudioWorkletProcessor and call their exported functions from it and it JustWorks(TM).

I've been going through this code back and forth for a few passes now, and some of the design questions on this front do boil down to the portability vs native web question. You mention portability, but there seems to be currently very little provided in terms of portability in the test code, since it is very JS-specific?

The review and maintenance complexity of this PR is high, so we should be very careful about how to proceed. Some thoughts to get us further:

- the test code shows an example of how to generate an audio tone. It is not immediately clear to me what is intended to be the public user-facing API of the feature, and what is intended to be part of the test code only. That is, if I wanted to use AudioWorklet pthreads in my own code, what API should I work against?

Maybe we'll need to flesh that out in documentation or a hello world test without test harness, to illustrate the expected API surface area.

- the other question is about portability vs native web: you mention portability being a goal, but it looks like the test code contains an example of a JS-specific code API that needs to be called to create a "AudioWorklet pthread". That kind of API does not seem to actually aim for portability, since JS code is very JS specific?

My expectation for portability would be something akin to having a C based API, where one would call something like `audio_worklet_pthread_create()` instead of `pthread_create()` to create a pthread that runs inside an AudioWorklet, and the thread startup routine would be the audio buffer mixing routine - or something like that?

- Then, there is a question of code organization. It seems to me that some of the work could be located in a `library_audioworklet.js` instead, rather than piling it in the `library_pthread.js` API. Also, the code test example feels somewhat odd as an API, but I am not able at the moment to suggest how this should look like instead - I should probably give this an in-depth playthrough to get a feeling.

- Finally, code size is a concern. We will need to land this in a fashion that does not regress code size _a single byte_ when one is building without `worklet` enabled in the build settings. Currently it looks like all projects will pay a code size hit from AudioWorklet related code even though they would not be interested in the feature. That is something we'll need to be very precise about.

To summarize, in order to proceed with this feature, I believe we should define in detail, what the "native web" API would look like to approach this feature, if portability to other platforms is not a concern.

Then we can define what the "portable" API perspective would look like, if/when the added requirement of portability is necessary - and how should that API differ from the native web-like API?

tklajnscek commented:
@juj I'll try to find some time to work on this in the next few days and we can discuss. For now here's a quick reply to your points.

> Maybe we'll need to flesh that out in documentation or a hello world test without test harness, to illustrate the expected API surface area.

Yeah, I think a clean sample would make sense. I'll put something somewhere that we can use a starting point for discussion.

> the other question is about portability vs native web: you mention portability being a goal, but it looks like the test code contains an example of a JS-specific code API that needs to be called to create a "AudioWorklet pthread". That kind of API does not seem to actually aim for portability, since JS code is very JS specific?

The _port existing code_ part of my comments above was referring to the ability to easily take existing C(++) audio _processing_ code and make it work pretty much as-is under Emscripten when executed in an audio worklet context.  My intention was not to create a generic, portable API for access to the platform-specific audio system which this would become. I didn't even attempt to wrap/abstract/hide the creation the AudioContext etc which immediately makes this a non-portable thing. I think that is totally beyond the scope of Emscripten and it would most likely be a total nightmare to maintain.

What I was aiming for was to add a way of hooking into the audio callbacks like you can on other platforms (using platform specific APIs - XAudio/WASAPI/AU etc.) and use normal C++ code in there without having to modify the internals of Emscripten/pthreads/low level synchronization. This is what I was going for with portability. Because currently, you can't write a function that gets called from an audio worklet and actually any pthread/sync primitive at all.

For example, if you have an existing app that renders audio you likely have a bunch of platform-independent code that processes all the audio/dsp logic. That's the bulk of the code. Then you have some platform-specific code that queries supported features/formats, creates the audio device and ends upsetting up some kind of DPC audio callback which is where the audio gets submitted to the os/driver for playback. And that should stay like that IMO.

Before AudioWorklets this worked fine in Emscripten too. For the 'Emscripten platform' part of the code we used a ScriptAudioProcessor node and simply called our "native" DPC callback to fill the buffers from `onaudioprocess`. It worked, because it all executed in the same context as the main code on the main thread. But as soon as you move this off the main thread you need some kind of atomics/sync logic and you need Emscripten compiled code _to function as intended_ when executed from within the audio worklet context (which is a bit special). This is what this PR is trying to address. To just add the ability to even run the code from the audio worklet as if it were just another normal thread.

> * It seems to me that some of the work could be located in a `library_audioworklet.js` instead, rather than piling it in the `library_pthread.js`

Yeah, definitely. I'll make a pass on this to clean it up.

> * Finally, code size is a concern. We will need to land this in a fashion that does not regress code size _a single byte_ when one is building without `worklet` enabled in the build settings

While I totally agree with the sentiment, this is more of a question of balance between code duplication and keeping code size down when disabled. @kripken asked me to deduplicate some of the code, which I also think is important to make it maintainable. But at the same time, doing this means that it because more difficult to keep the overhead to an absolute 0 when disabled without making the whole thing a hot ifdef-ridden mess. It becomes especially difficult in the futex implementation since just extracting a part of it into a common function _will increase_ the code size even for non-audio-worklet builds.

In any case, I'm confident we can get pretty damn close to 0, or you know, optimize something else in the process to offset the cost :)

tklajnscek committed:
Support running Emscripten-compiled code from AudioWorklets (almost) as if they were regular Worker pthreads

tklajnscek commented:
Hey everyone,

So I said _in a few days_ but then life happened, so here we are 6 months later and I've finally been able to spend some more time on this and brought it up to speed with the latest revision and addressed some of the issues, mostly:

- Made sure that code size doesn't change unless you actually enable the audio worklet environment
- `tests/audioworklet/audioworklet.cpp` is now cleanly separated into _test-only_ and _usage example_ parts where literally just deleting all the test-only parts yields a working usage example.
- I didn't end up splitting the audio worklet code in `library_pthread.js` out into `library_audioworklet.js` as @juj suggested because it's so intimately tied with all the pthread stuff (it's essential just a different way to set it up and then it all just works)

I'd appreciate any guidance on what needs doing so that I stand a chance of getting this merged as I really thing a lot of cool stuff can be done in the audio field with Emscripten and this just brings it way closer to native audio dev then it used to be.

_Note: Some of the SSE tests seem to be failing, but I'm not sure how this is related to my changes and it seems to be non-trivial for me to run them locally, but I can try if this is not a known issue with the test suite?_

sbc100 commented on src/settings_internal.js:
```diff
@@ -139,6 +139,7 @@ var ENVIRONMENT_MAY_BE_WORKER = 1;
 var ENVIRONMENT_MAY_BE_NODE = 1;
 var ENVIRONMENT_MAY_BE_SHELL = 1;
 var ENVIRONMENT_MAY_BE_WEBVIEW = 1;
+var ENVIRONMENT_MAY_BE_AUDIOWORKLET = 1;
```
Oh I see the setup code unconditionally sets its so the default is not particularly relevant.  I wonder if we should change that so the defaults listed here are meanful?  (In which change 0 would make more sense here).   But that can be a followup PR maybe?

tklajnscek replied:
Yeah, I guess this needs a bit of cleanup as a whole so I agree deferring this to a separate PR is better.

sbc100 commented on src/shell.js:
```diff
@@ -343,6 +346,8 @@ if (ENVIRONMENT_IS_WEB || ENVIRONMENT_IS_WORKER) {
   }
 
   setWindowTitle = function(title) { document.title = title };
+} else if (ENVIRONMENT_IS_AUDIOWORKLET) {
```
What these new lines in `#if ENVIRONMENT_MAY_BE_AUDIOWORKLET`?

sbc100 commented on src/shell_minimal.js:
```diff
@@ -138,7 +138,9 @@ var _scriptDir = (typeof document !== 'undefined' && document.currentScript) ? d
-var ENVIRONMENT_IS_WORKER = ENVIRONMENT_IS_PTHREAD = typeof importScripts === 'function';
+var ENVIRONMENT_IS_WORKER = typeof importScripts === 'function';
+var ENVIRONMENT_IS_AUDIOWORKLET = typeof AudioWorkletGlobalScope === 'function';
+var ENVIRONMENT_IS_PTHREAD = ENVIRONMENT_IS_WORKER || ENVIRONMENT_IS_AUDIOWORKLET;
```
Again, we don't want to pay the codesize hit here unless `ENVIRONMENT_MAY_BE_AUDIOWORKLET`

sbc100 commented on src/worker.js:
```diff
@@ -175,13 +218,20 @@ self.onmessage = function(e) {
 #else
+#if ENVIRONMENT_MAY_BE_AUDIOWORKLET
+      // When running as an AudioWorklet all the scripts are imported from the main thread (via .addModule)      
+      if(!{{{ makeAsmImportsAccessInPthread('ENVIRONMENT_IS_AUDIOWORKLET') }}}) { 
```
space after `if`

sbc100 commented on tests/audioworklet/audioworklet.cpp:
```diff
@@ -0,0 +1,190 @@
+  invSampleRate = 1.0f / sampleRate;
+  printf("Initialized audio context. Sample rate: %d. PThread init pending.\n", sampleRate);
+  createAudioWorklet();
+  printf("Creating audio worklet.\n");
```
Shouldn't this print come before the call to `createAudioWorklet` (or be phrased in the past tense).

sbc100 commented on tests/audioworklet/audioworklet.cpp:
```diff
@@ -0,0 +1,190 @@
+  printf("Initialized audio context. Sample rate: %d. PThread init pending.\n", sampleRate);
+  createAudioWorklet();
+  printf("Creating audio worklet.\n");
+}
```
return 0;

sbc100 commented on tests/audioworklet/audioworklet_post.js:
```diff
@@ -0,0 +1,47 @@
+
+  // Register the processor as per the audio worklet spec
+  registerProcessor('native-passthrough-processor', NativePassthroughProcessor);
+}
```
missing newline

sbc100 commented on tests/audioworklet/shell.html:
```diff
@@ -0,0 +1,1305 @@
+    </script>
+  </body>
+</html>
+
```
Do we need to check in this huge html file?   Where does it come from?

tklajnscek replied:
It's just a copy of `src/shell.html` customized with the export name because:
```
emcc: error: pthreads + MODULARIZE currently require you to set -s EXPORT_NAME=Something (see settings.js) to Something != Module, so that the .worker.js file can work
```

And then setting EXPORT_NAME
```
emcc: error: Customizing EXPORT_NAME requires that the HTML be customized to use that name (see https://github.com/emscripten-core/emscripten/issues/10086)
```

Actually, if I make it run with MINIMAL_RUNTIME I can get rid of this... hold on :)

tklajnscek replied:
Couldn't make it run with MINIMAL_RUNTIME so far, but I did find that other tests use `test_file('shell_that_launches_modularize.html')` that loads `MyModule` so I just used that as well and renamed my module to MyModule! 

So this removes the entire html.

sbc100 commented on tests/test_browser.py:
```diff
@@ -5103,6 +5103,17 @@ def test_assert_failure(self):
+  def test_audio_worklet(self):
+    self.btest(path_from_root('tests', 'audioworklet', 'audioworklet.cpp'),
+               expected='1',
+               args=['-s', 'USE_PTHREADS=1', '-s', 'MODULARIZE=1', '-s',
```
Is `MODULARIZE` needed to use worklets?

tklajnscek replied:
Yes, unfortunately it's needed because AudioWorkletGlobalScope doesn't let us fetch/import scripts from the worklet context which is what regular pthreads do (they load the `.worker.js` which then does `importScript` in the `load` message from the main thread). 

In the worklet case we need to use `audioCtx.addModule` on the main thread to load both the `.worker.js` and the main js and then when they're both loaded and the `load` message is sent over we can just skip the import step and run the module function. 

If we don't MODULARIZE then then the main js executes before all the things are set up in the `load` message coming from the main thread (the `wasmModule` etc). For the regular pthread case this is not a problem, because `importScripts` happens in `load` only after this has been set up already.

tklajnscek replied:
I actually found a way to make it work without MODULARIZE, so this is resolved now.

sbc100 commented on tests/test_browser.py:
```diff
@@ -5103,6 +5103,17 @@ def test_assert_failure(self):
+  @requires_sound_hardware
+  def test_audio_worklet(self):
+    self.btest(path_from_root('tests', 'audioworklet', 'audioworklet.cpp'),
+               expected='1',
```
I wonder if you can use `btest_exit` here, and use `exit(0)` to signal success in your test (I'm trying to move browser tests in that direction in general).

tklajnscek replied:
I don't think I can, because setting EXIT_RUNTIME will exit the runtime when `main()` returns, but I need to actually wait for the AudioContext to start, then for the worklet runs and do it's thing.

sbc100 commented on tests/audioworklet/audioworklet.cpp:
```diff
@@ -0,0 +1,190 @@
+
+EMSCRIPTEN_KEEPALIVE extern "C" void beforeUnload() {
+  // Always release the mutex on unload, otherwise the worklet can get stuck if we navigate away
+  testMutex.unlock();
```
But won't the worklet get killed in this case?   Is it really possible to "leak" a worklet when you navigate away?     This is not true for normal pthreads/workers.

tklajnscek replied:
This test is testing the worklet futex support so in `signalTestSuccess` the worklet tries to lock the mutex and it starts waiting for the main thread. If the main thread never makes it to the code that unlocks the mutex (e.g. you kill it before by navigating away) the worklet remains stuck in a loop and the tab becomes unresponsive. 

So it won't _leak_ the worklet, it'll try to clean it up and it'll become unresponsive :)

Workers don't do this, because they use `Atomics.Wait/Notify` to wait on a mutex which is non-blocking, but if the main thread was waiting on a mutex and the worker that's holding it crashes and doesn't unlock it then the same thing will happen (tab unresponsive)

sbc100 reviewed (commented):
In general, I'd love to see some the complexity of the test/example reduced.   It would seem odd to be if that was the smallest possible test/example of using a worklet.

tklajnscek commented:
Wow, thanks for the uber fast review! I'll jump on it first thing in the morning as it's quite late here already :)

tklajnscek commented on emcc.py:
```diff
@@ -3248,7 +3251,17 @@ def modularize():
+
+    # Store the export on the global scope for audio worklets
+    if settings.ENVIRONMENT_MAY_BE_AUDIOWORKLET:
+      f.write('''if (typeof AudioWorkletGlobalScope === 'function')
```
It's where all the exports are registered (check just below this) - I think it's because the EXPORT_NAME gets replaced this can't be in `postamble.js`.

And yes, good call on fstrings, but the rest of the module registration code below doesn't use them so I just went with the same style. Will change it.

tklajnscek committed:
Addressed @sbc's comments
Split tests/audioworklet into a futex test (without the tone generator) and a tone generator example (without the test stuff)
Fixed `moduleLoaded` failing (it needs to register on the global scope)
Made everything build with MINIMAL_RUNTIME (doesn't run just yet)
Removed the acron fix temporarily

tklajnscek commented:
In addition the the stuff discussed in the diff comments I split up the worklet test into a futex test and a tone generator example.

I'll amend this PR some more after I'm done playing around with MODULARIZE=0/1 and MINIMAL _RUNTIME.. have some ideas :)

tklajnscek committed:
Made audio worklets work without MODULARIZE too (by loading the main JS via worklet.addModule on the main thread)
Removed some leftover debug code (registerProcessor override)

tklajnscek commented:
Let's see if the tests pass! Made everything work without requiring MODULARIZE=1! :)

tklajnscek committed:
Some final notes & cleanup
Updated ChangeLog.md

tklajnscek committed:
Updated change log

tklajnscek commented:
I think we're good to go here. Any further feedback is much appreciated!

tklajnscek committed:
tidy onmessage and make it work with closure

tklajnscek committed:
fixed non-modularized variant to work with --closure=1
cleaned up audio_worklet tests and made both non-modularized and moduliarized variants also test with --closure=1

tklajnscek committed:
make flake8 happy again

tklajnscek commented:
Some more fixes:
- Realized `MODULARIZE=0 --closure=1` was broken so fixed that and made the test actually test all these combinations.
- A few other `--closure=1` fixes (message fields accessed as strings mostly)
- Some minor reorg of the `onmessage` handler to remove traces of worklet specifics in the regular worker `onmessage`

I think this is shaping up nicely if I do say so myself 😁

tklajnscek commented:
Ping :) Has anyone had time to look at this?
I really think this is something that's generally useful and I'd like to get it merged.

tklajnscek commented:
This is my fortnightly reminder for an update here 😄 
It would be much appreciated 🙏

kripken commented:
I tried to do another reading of this, but given all the discussion, II'm not sure where things stand. In particular, I see there was discussion of a possible C API and other options, for example - are those discussions still ongoing? Or is this PR in an intended final state?

Skimming the code, it does look like it no longer duplicates a bunch of logic, which is good!

tklajnscek commented:
Regarding the C API, I still stand by what I said above - I don't think it would be very useful or add to portability in any meaningful way, because there's still JS that needs to be written for the AudioWorkletProcessor as well as the AudioContext itself. So I'd just consider this _platform-specific audio code_ which everyone will have either as part of their code or some audio middleware they use and they can do whatever needs doing for their use case there.

I haven't received any further comment on it so I just assumed I had agreement on this 😄 

Here's a longer take on this:

The current init code is very minimal and I think making it an even smaller C API would likely remove a bunch of options/control from the user... this is from the example:
```
  // Create the context
  Module.audioCtx = new AudioContext();

  // Initialize the pthread shared by all AudioWorkletNodes in this context
  PThread.initAudioWorkletPThread(Module.audioCtx, pthreadPtr).then(function() {
    out("Audio worklet PThread context initialized!")
  }, function(err) {
    out("Audio worklet PThread context initialization failed: " + [err, err.stack]);
  });

  // Creates an AudioWorkletNode and connects it to the output once it's created
  PThread.createAudioWorkletNode(
    Module.audioCtx,
    'native-passthrough-processor', 
    {
      numberOfInputs: 0,
      numberOfOutputs : 1,
      outputChannelCount : [2]
    }
  ).then(function(workletNode) {
    // Connect the worklet to the audio context output
    out("Audio worklet node created! Tap/click on the window if you don't hear audio!");
    workletNode.connect(Module.audioCtx.destination);
  }, function(err) {
    out("Audio worklet node creation failed: " + [err, err.stack]);
  });
```

If we ignore the AudioContext creation which the user will probably want to create themselves with their own specific options etc. (they're most likely doing this in some way already if they have audio playback either directly or SDL/OpenAL) we're left with the worklet pthread init and the processor creation and connection/disconnection. If we want to leave it flexible then we can do something like this for example

`int audio_worklet_pthread_create(pthread_t* thread, AUDIOCONTEXT);`
This would need the AudioContext JS object passed in somehow, which means it would have to register somehwere first so that the C API has an int handle to it. This is already pretty ugly IMO.

`int audio_worklet_node_create(pthread_t* thread, JSAudioContext, const char* processorName, void (*doneCallback)(int), OPTIONS)`
The problem with creating a node is that it takes [options](https://developer.mozilla.org/en-US/docs/Web/API/AudioWorkletNode/AudioWorkletNode) which are a nested JS object with custom fields (e.g. processorOptions) so this can't really be passed in C-style. It's also async so it needs a completion callback which is a bit meh.

So just this is pretty bad already and doesn't really gain us anything.

The other option would be to create a much higher level C API that abstracts all of this away and it just becomes something like:
`int audio_worklet_pthread_create(pthread_t* thread, AUDIOCONTEXT, void (*audioCallback)(float* data, int numSamples), int numChannels);`

Which still has the issue of how to pass in the AudioContext and really removes most of the user control which I think makes it useless beyond very basic use cases and severely limits the usefulness of this whole PR.

*TL;DR:* I think the PR like it is adds the ability to properly execute native Emscripten code in AudioWorklet scope and gives you a few simple JS methods to get it all set up easily without limiting what you can do with it too much. I consider it good to go. Well, except the merge conflicts that have popped up since :P

kripken commented:
I see, thanks @tklajnscek !

To me that sounds pretty reasonable. However, we should probably wait for @juj in order to make a decision on the direction, as the other person that has looked into this and would like to use it. @juj may not be available to respond here for some weeks, though, so that means a significant delay. I don't see a way around that, however - if the code were simpler, I'd suggest maybe landing soon and revising or replacing later, but even after the large simplifications here, that might not be easy to do.

So I think this might need to wait for a full discussion on it. Sorry about that - this is useful functionality and I appreciate the work here!

tklajnscek commented:
No problem @kripken, we can wait. I guess this won't code-rot too much in a few weeks time :)

Looking forward to discussing this. We can also take this to discord if it makes sense, just let me know when @juj is ready to go.

I'll ping back in a few weeks to remind everyone.

kripken commented:
ping @juj - thoughts?

cwoffenden commented:
Since Chrome 93 this PR (or anything with AudioWorklets and wasm) also requires this LLVM patch:

https://bugs.llvm.org/show_bug.cgi?id=51702

juj commented:
Heya, sorry for the delay on this, I was on a leave this summer. Just adding a note here to ack that I've been looking at this now. I'll post a more thorough writeup on this soon.

tklajnscek committed:
Implemented a Chrome v95+ bug (?) workaround to pass the initial load message to the audio worklet dummy processor in `processorOptions` instead of as a regular `port.sendMessage` which doesn't get delievered anymore if it contains a WebAssembly.Module

Jonathhhan commented:
That sounds really great. I use Open Frameworks with Emscripten, and it would be very nice to process audio with an audioworklet instead of scriptprocessornode. I can already use the sharedarraybuffer from the wasm Module in the audioworklet (I only need to send it once), but since I send messages for the audiocallback to the main thread it it less efficient than before. I would be really great to use the audiocallback from the wasm module in the audioworklet. sadly i couldnt compile with your branches (i use the latest emsdk).

tklajnscek committed:
fixed for 2.0.34

tklajnscek commented:
Hey @Jonathhhan, yeah I haven't been merging in the latest changes lately... I've gone ahead and rebased everything on 2.0.34 so this should work for you:

1. Clone the base SDK folder from `https://github.com/emscripten-core/emsdk.git`
2. `emsdk install 2.0.34`
3. `emsdk activate 2.0.34`
4. remove the content of `<sdkdir>/upstream/emscripten`
5. clone the `worklet_support` branch from my repo `https://github.com/tklajnscek/emscripten.git` in there

This should do it. Then go into `tests` and try running `runner.py browser.test_audio_worklet` and it should succeed.

The tone generator sample should also build and run - in `tests/audioworklet/tone` - check the .cpp file for instructions.

Let me know if it works for you :)
Also, be aware that it doesn't look likely this will ever make it into Emscripten in it's current form :)

tklajnscek committed:
fixed module initalization after rebase

Jonathhhan commented:
@tklajnscek Thank you very much - I was looking for a solution like that for a long time. Open Frameworks is working with audioworklet now. I make an audiocallback from the audioworklet every 128 samples. A strange thing is that it only runs smooth with Chrome and Ubuntu, other browser / OS combinations have some audio artifacts / glitches - maybe the buffersize is too small, I will investigate on that. Here is my example: https://simplesequencer.handmadeproductions.de/ Here is the same thing without audioworklet (and a buffersize of 4096): https://simplesequencer.handmadeproductions.de/

Jonathhhan commented:
I set the Open Frameworks buffersize to 2048 and it is much better now.
This is how I implemented the code (I hope it is not too off-topic):
`	html5audio_stream_create: function(context_id, bufferSize, inputChannels, outputChannels, inbuffer, outbuffer, callback, userData, pthreadPtr){
// Initializes the audio context and the pthread it it's AudioWorkletGlobalScope

  // Create the context
  Module.audioCtx = new AudioContext({sampleRate: 88200});

  // Initialize the pthread shared by all AudioWorkletNodes in this context
  PThread.initAudioWorkletPThread(Module.audioCtx, pthreadPtr).then(function() {
    console.log("Audio worklet PThread context initialized!")
  }, function(err) {
    console.log("Audio worklet PThread context initialization failed: " + [err, err.stack]);
  });

  // Creates an AudioWorkletNode and connects it to the output once it's created
  PThread.createAudioWorkletNode(
    Module.audioCtx,
    'native-passthrough-processor', 
    {
      numberOfInputs: 0,
      numberOfOutputs : 1,
      outputChannelCount : [2],
      processorOptions: {
      inputChannels : inputChannels,
      outputChannels : outputChannels,
      inbuffer : inbuffer,
      outbuffer : outbuffer,
      bufferSize : bufferSize,
      callback : callback,
      userData : userData
      }     
    }
  ).then(function(workletNode) {
    // Connect the worklet to the audio context output
    console.log("Audio worklet node created! Tap/click on the window if you don't hear audio!");
    workletNode.connect(Module.audioCtx.destination);
  }, function(err) {
    console.log("Audio worklet node creation failed: " + [err, err.stack]);
  });

  // To make this example usable we setup a resume on user interaction as browsers
  // all require the user to interact with the page before letting audio play
  if (window && window.addEventListener) {
    var opts = { capture: true, passive : true };    
    window.addEventListener("touchstart", function() { Module.audioCtx.resume() }, opts);
    window.addEventListener("mousedown", function() { Module.audioCtx.resume() }, opts);
    window.addEventListener("keydown", function() { Module.audioCtx.resume() }, opts);
  }
	},

	html5audio_stream_free: function(stream_id){
		return AUDIO.streams[stream_id] = null;
		return AUDIO.mediaElements[stream_id] = null;
	},`

This is the audioworklet:

```
/**
 * This is the JS side of the AudioWorklet processing that creates our
 * AudioWorkletProcessor that fetches the audio data from native code and 
 * copies it into the output buffers.
 * 
 * This is intentionally not made part of Emscripten AudioWorklet integration
 * because apps will usually want to a lot of control here (formats, channels, 
 * additional processors etc.)
 */

// Register our audio processors if the code loads in an AudioWorkletGlobalScope
if (typeof AudioWorkletGlobalScope === "function") {
  // This processor node is a simple proxy to the audio generator in native code.
  // It calls the native function then copies the samples into the output buffer
  var counter = 0;
  var inputChannels = 0;
  var outputChannels = 0;
  var inbuffer = 0;
  var outbuffer = 0;
  var bufferSize = 0;
  var callback = 0;
  var userData = 0;
  class NativePassthroughProcessor extends AudioWorkletProcessor {
    constructor (options) {
    super()
    inputChannels = options.processorOptions.inputChannels;
    outputChannels = options.processorOptions.outputChannels;
    inbuffer = options.processorOptions.inbuffer;
    outbuffer = options.processorOptions.outbuffer;
    bufferSize = options.processorOptions.bufferSize;
    callback = options.processorOptions.callback;
    userData = options.processorOptions.userData;
  }
    process(inputs, outputs, parameters) {
    counter = currentFrame / 128 % 32;
    if (counter == 0) {
        dynCall('viiii',callback, [bufferSize,inputChannels,outputChannels,userData]);
    }
    const output = outputs[0];
    for (let channel = 0; channel < 2; ++channel) {
        const outputChannel = output[channel];
        outputChannel.set(Module.HEAPF32.subarray(outbuffer>>2,(outbuffer>>2)+bufferSize*outputChannels).slice(counter * 128, counter * 128 + 128));
        }
        return true;
    }
}
  // Register the processor as per the audio worklet spec
  registerProcessor('native-passthrough-processor', NativePassthroughProcessor);
}
```
My only issue is that I need to edit the generated java script file, because I do not know how to change the source code for that.
Does somebody know how to change the source code for that?
If I use files and a filesystem (the problem is cause by `var xhr = new XMLHttpRequest;`) I nees to set it in an `if (ENVIRONMENT_IS_WEB){}` condition.
My change (its the beginning of the generated file) looks like that:
`if (typeof AudioWorkletGlobalScope === "function") {
    var Module = globalThis.Module
}
var Module = typeof Module !== "undefined" ? Module : {};
if (!Module.expectedDataFileDownloads) {
    Module.expectedDataFileDownloads = 0
}
Module.expectedDataFileDownloads++;
var ENVIRONMENT_IS_WEB = typeof window === "object";
var ENVIRONMENT_IS_WORKER = typeof importScripts === "function";
var ENVIRONMENT_IS_AUDIOWORKLET = typeof AudioWorkletGlobalScope === "function";
var ENVIRONMENT_IS_NODE = typeof process === "object" && typeof process.versions === "object" && typeof process.versions.node === "string";
var ENVIRONMENT_IS_PTHREAD = Module["ENVIRONMENT_IS_PTHREAD"] || false;
if (ENVIRONMENT_IS_WEB){
(function() {
    var loadPackage = function(metadata) {
        var PACKAGE_PATH = "";
        if (typeof window === "object") {
            PACKAGE_PATH = window["encodeURIComponent"](window.location.pathname.toString().substring(0, window.location.pathname.toString().lastIndexOf("/")) + "/")
        } else if (typeof process === "undefined" && typeof location !== "undefined") {
            PACKAGE_PATH = encodeURIComponent(location.pathname.toString().substring(0, location.pathname.toString().lastIndexOf("/")) + "/")
        }
        var PACKAGE_NAME = "bin/EmscriptenExample.data";
        var REMOTE_PACKAGE_BASE = "EmscriptenExample.data";
        if (typeof Module["locateFilePackage"] === "function" && !Module["locateFile"]) {
            Module["locateFile"] = Module["locateFilePackage"];
            err("warning: you defined Module.locateFilePackage, that has been renamed to Module.locateFile (using your locateFilePackage for now)")
        }
        var REMOTE_PACKAGE_NAME = Module["locateFile"] ? Module["locateFile"](REMOTE_PACKAGE_BASE, "") : REMOTE_PACKAGE_BASE;
        var REMOTE_PACKAGE_SIZE = metadata["remote_package_size"];
        var PACKAGE_UUID = metadata["package_uuid"];
        function fetchRemotePackage(packageName, packageSize, callback, errback) {
            if (typeof process === "object" && typeof process.versions === "object" && typeof process.versions.node === "string") {
                require("fs").readFile(packageName, function(err, contents) {
                    if (err) {
                        errback(err)
                    } else {
                        callback(contents.buffer)
                    }
                });
                return
            }
            var xhr = new XMLHttpRequest;
            xhr.open("GET", packageName, true);
            xhr.responseType = "arraybuffer";
            xhr.onprogress = function(event) {
                var url = packageName;
                var size = packageSize;
                if (event.total)
                    size = event.total;
                if (event.loaded) {
                    if (!xhr.addedTotal) {
                        xhr.addedTotal = true;
                        if (!Module.dataFileDownloads)
                            Module.dataFileDownloads = {};
                        Module.dataFileDownloads[url] = {
                            loaded: event.loaded,
                            total: size
                        }
                    } else {
                        Module.dataFileDownloads[url].loaded = event.loaded
                    }
                    var total = 0;
                    var loaded = 0;
                    var num = 0;
                    for (var download in Module.dataFileDownloads) {
                        var data = Module.dataFileDownloads[download];
                        total += data.total;
                        loaded += data.loaded;
                        num++
                    }
                    total = Math.ceil(total * Module.expectedDataFileDownloads / num);
                    if (Module["setStatus"])
                        Module["setStatus"]("Downloading data... (" + loaded + "/" + total + ")")
                } else if (!Module.dataFileDownloads) {
                    if (Module["setStatus"])
                        Module["setStatus"]("Downloading data...")
                }
            }
            ;
            xhr.onerror = function(event) {
                throw new Error("NetworkError for: " + packageName)
            }
            ;
            xhr.onload = function(event) {
                if (xhr.status == 200 || xhr.status == 304 || xhr.status == 206 || xhr.status == 0 && xhr.response) {
                    var packageData = xhr.response;
                    callback(packageData)
                } else {
                    throw new Error(xhr.statusText + " : " + xhr.responseURL)
                }
            }
            ;
            xhr.send(null)
        }
        function handleError(error) {
            console.error("package error:", error)
        }
        var fetchedCallback = null;
        var fetched = Module["getPreloadedPackage"] ? Module["getPreloadedPackage"](REMOTE_PACKAGE_NAME, REMOTE_PACKAGE_SIZE) : null;
        if (!fetched)
            fetchRemotePackage(REMOTE_PACKAGE_NAME, REMOTE_PACKAGE_SIZE, function(data) {
                if (fetchedCallback) {
                    fetchedCallback(data);
                    fetchedCallback = null
                } else {
                    fetched = data
                }
            }, handleError);
        function runWithFS() {
            function assert(check, msg) {
                if (!check)
                    throw msg + (new Error).stack
            }
            Module["FS_createPath"]("/", "data", true, true);
            Module["FS_createPath"]("/data", "pd", true, true);
            Module["FS_createPath"]("/data/pd", "data", true, true);
            function DataRequest(start, end, audio) {
                this.start = start;
                this.end = end;
                this.audio = audio
            }
            DataRequest.prototype = {
                requests: {},
                open: function(mode, name) {
                    this.name = name;
                    this.requests[name] = this;
                    Module["addRunDependency"]("fp " + this.name)
                },
                send: function() {},
                onload: function() {
                    var byteArray = this.byteArray.subarray(this.start, this.end);
                    this.finish(byteArray)
                },
                finish: function(byteArray) {
                    var that = this;
                    Module["FS_createDataFile"](this.name, null, byteArray, true, true, true);
                    Module["removeRunDependency"]("fp " + that.name);
                    this.requests[this.name] = null
                }
            };
            var files = metadata["files"];
            for (var i = 0; i < files.length; ++i) {
                new DataRequest(files[i]["start"],files[i]["end"],files[i]["audio"] || 0).open("GET", files[i]["filename"])
            }
            function processPackageData(arrayBuffer) {
                assert(arrayBuffer, "Loading data file failed.");
                assert(arrayBuffer instanceof ArrayBuffer, "bad input to processPackageData");
                var byteArray = new Uint8Array(arrayBuffer);
                DataRequest.prototype.byteArray = byteArray;
                var files = metadata["files"];
                for (var i = 0; i < files.length; ++i) {
                    DataRequest.prototype.requests[files[i].filename].onload()
                }
                Module["removeRunDependency"]("datafile_bin/EmscriptenExample.data")
            }
            Module["addRunDependency"]("datafile_bin/EmscriptenExample.data");
            if (!Module.preloadResults)
                Module.preloadResults = {};
            Module.preloadResults[PACKAGE_NAME] = {
                fromCache: false
            };
            if (fetched) {
                processPackageData(fetched);
                fetched = null
            } else {
                fetchedCallback = processPackageData
            }
        }
        if (Module["calledRun"]) {
            runWithFS()
        } else {
            if (!Module["preRun"])
                Module["preRun"] = [];
            Module["preRun"].push(runWithFS)
        }
    };
    loadPackage({
        "files": [{
            "filename": "/data/audioworklet_tone_post.js",
            "start": 0,
            "end": 1469
        }, {
            "filename": "/data/pd/ofelia_bang.pd",
            "start": 1469,
            "end": 2894
        }, {
            "filename": "/data/pd/ofelia_vSlider.pd",
            "start": 2894,
            "end": 5102
        }, {
            "filename": "/data/pd/ofelia_hRadio.pd",
            "start": 5102,
            "end": 6725
        }, {
            "filename": "/data/pd/ofelia_hSlider.pd",
            "start": 6725,
            "end": 8917
        }, {
            "filename": "/data/pd/ofelia_label.pd",
            "start": 8917,
            "end": 9801
        }, {
            "filename": "/data/pd/ofelia_vRadio.pd",
            "start": 9801,
            "end": 11435
        }, {
            "filename": "/data/pd/ofSetup.lua",
            "start": 11435,
            "end": 12178
        }, {
            "filename": "/data/pd/main.pd",
            "start": 12178,
            "end": 63469
        }, {
            "filename": "/data/pd/gadsr.pd",
            "start": 63469,
            "end": 66878
        }, {
            "filename": "/data/pd/ofSetup.pd",
            "start": 66878,
            "end": 67142
        }, {
            "filename": "/data/pd/preset.pd",
            "start": 67142,
            "end": 68577
        }, {
            "filename": "/data/pd/ofelia_number.pd",
            "start": 68577,
            "end": 72170
        }, {
            "filename": "/data/pd/ofelia_toggle.pd",
            "start": 72170,
            "end": 73626
        }, {
            "filename": "/data/pd/synthvoice.pd",
            "start": 73626,
            "end": 75200
        }, {
            "filename": "/data/pd/data/Raster.frag",
            "start": 75200,
            "end": 75827
        }, {
            "filename": "/data/pd/data/Raster.vert",
            "start": 75827,
            "end": 76079
        }],
        "remote_package_size": 76079,
        "package_uuid": "71d3d0c5-3473-469b-9e09-d610538b347d"
    })
}
)();`

tklajnscek commented:
Hey @Jonathhhan, I tried your demos and they both actually work fine for me in Chrome on Win 10 - couldn't hear any glitches :)

Obviously you have to be careful with any code you run in the audio worklet processor as it has to be very fast to stay real-time. Also, as you already noticed, the buffer is settable and that helps, but as far as I understand, the browser is always supposed to figure the ideal buffer to use with audio worklets and that's the whole point of a fixed 128 sample quantum.

If you can't make it work fast enough, then you might have to run the processing in another (multiple?) thread(s) and then just use the audio worklet processor to fill the output buffers.

Finally, regarding the file packager thingy - we don't actually use that so I'm not going to be able to help you much there, but generally, the issue is that XHRs are not allowed in worklet contexts. 

The first solution is to fix `file_packager.py` to generate code that skips itself like you're doing now manually every time (`if(!ENVIRONMENT_IS_AUDIOWORKLET)` should be enough btw) .

The other, IMO better, option is switch to manual file fetches which are more flexible, but it requires you to know in the code that you need to download stuff and wait for it before using file.
Roughly, what you need to do is:

1. Mount a file system (e.g. `FS.mount(MEMFS...`)
2. Then call `emscripten_async_wget2` to download files into it
3. Once #2 is done, use the file as usual from the path you downloaded it into

Hope any of this helps :)

Jonathhhan commented:
@tklajnscek
thank you very much for the hints (will try to optimize the patch an the implementation).
Actually it works also well with a buffersize of 512, with 128 it depends on the browser and OS. All in all it is a big improvement and makes audio in the web reliable (basically I run Pure Data patches with OF and Emscripten)...

Jonathhhan commented:
Hey @tklajnscek, do you maby know if it is possible to run audioworklets with your branch together with -s PROXY_TO_PTHREAD=1 -s OFFSCREENCANVAS_SUPPORT=1 -s OFFSCREEN_FRAMEBUFFER=1? I tried, but get this error message:
```
EmscriptenExample.worker.js:1 worker.js onmessage() captured an uncaught exception: TypeError: Cannot read properties of undefined (reading 'audioWorklet')
self.onmessage @ EmscriptenExample.worker.js:1
EmscriptenExample.worker.js:1 TypeError: Cannot read properties of undefined (reading 'audioWorklet')
    at Object.initAudioWorkletPThread (EmscriptenExample.js:1484:27)
    at _html5audio_stream_create (EmscriptenExample.js:9687:13)
    at ofxEmscriptenSoundStream::setup(ofSoundStreamSettings const&) (EmscriptenExample.wasm:0x21b75)
    at ofxOfelia::init(int, int, int, int, bool, bool, int, int, bool, std::__2::basic_string<char, std::__2::char_traits<char>, std::__2::allocator<char> > const&) (EmscriptenExample.wasm:0x28ae6)
    at ofApp::setup() (EmscriptenExample.wasm:0x1d57a)
    at ofNode::onParentOrientationChanged(glm::qua<float, (glm::qualifier)0>&) (EmscriptenExample.wasm:0x523af3)
    at std::__2::__function::__func<std::__2::shared_ptr<of::priv::Function<ofEventArgs, std::__2::recursive_mutex> > ofEvent<ofEventArgs, std::__2::recursive_mutex>::make_function<ofEasyCam>(ofEasyCam*, void (ofEasyCam::*)(ofEventArgs&), int)::'lambda'(void const*, ofEventArgs&), std::__2::allocator<std::__2::shared_ptr<of::priv::Function<ofEventArgs, std::__2::recursive_mutex> > ofEvent<ofEventArgs, std::__2::recursive_mutex>::make_function<ofEasyCam>(ofEasyCam*, void (ofEasyCam::*)(ofEventArgs&), int)::'lambda'(void const*, ofEventArgs&)>, bool (void const*, ofEventArgs&)>::operator()(void const*&&, ofEventArgs&) (EmscriptenExample.wasm:0x51c5d6)
    at ofEvent<ofHttpResponse, std::__2::recursive_mutex>::notify(ofHttpResponse&) (EmscriptenExample.wasm:0x2076c)
    at ofCoreEvents::notifySetup() (EmscriptenExample.wasm:0x533088)
    at ofxAppEmscriptenWindow::loop() (EmscriptenExample.wasm:0x1f6ab)
```
https://github.com/openframeworks/openFrameworks/issues/6781

tklajnscek commented:
Ha! It probably doesn't work because it needs code adding to proxy the call to the main thread, just like the regular pthread_create call does. There's also special case code for offscreen canvas support.

Since I don't really have a use case for this and I have other priorities right now I won't really be able to help you out with this, but you're more than welcome to take my code and make any changes necessary to make it work. I guess you could just look at the proxying and offscreen canvas code paths in `__pthread_create_js` and follow the same principles?

Jonathhhan commented:
@tklajnscek I wanted to thank you again, your implementation works absolutely great with Open Frameworks and Pure Data (I think I fixed all the isues I had). I build a small demo where you can manipulate local audio files and audio input: https://emscriptenpdeffect.handmadeproductions.de/

tklajnscek commented:
@Jonathhhan that's really great! I'm happy this was useful to someone else! 
Did you have to make any further changes to the internals of emscripten (and my changes) to make it work?

Jonathhhan commented:
@tklajnscek actually even the filesystem error disappeared after removing -s ASSERTIONS (which seems a bit strange, but anyhow). I made one small change for recalculating mouse coordinates in fullscreen mode and one for offscreen canvas (but that has nothing to do with the audioWorklet implementation).

Jonathhhan commented:
Hey @tklajnscek,
I am just courious if you have any plans to update your branch for future Emscripten versions?
If not, would it be difficult to update it myself? Not that it is necessary at the moment, but it would be great if audioWorklet and Emscripten will still work in the future.

tommitytom commented:
I'd just like to add that I'm using this branch for one of my projects and it has been absolutely fantastic so far.  The project is relatively complex and shares a lot of data between the main/audio thread via the use of the moodycamel concurrent queues and it works just as I expect it to!

kripken commented:
From the discussion above I think the general agreement was to focus on @juj 's Wasm Workers PR and to add AudioWorklet support on top of that? If so, then helping in that direction would probably be better. Or is there not consensus on that direction?

tklajnscek commented:
Oooh! I see there some action over at https://github.com/emscripten-core/emscripten/pull/12833 now! I was afraid it was dead after a few months of silence and was just about to start asking what's up :)

Also, I'm glad someone else is finding the work here useful! 🥳 

Anyways, my plan for this PR is to keep it reasonably updated to at least cover our needs for Emscripten updates until we hopefully end up building our thing on top of the web workers API. We usually take a new version every 3-4 months.

For our code base it shouldn't be a huge problem since we already have a per-platform threading abstraction so this is a natural fit where I can plug in the wasm workers API.

I guess this might not be that straightforward for others that have pthreads code sprinkled all over their codebase...

It'll be a few months until I get around to actually trying the wasm workers though as we're in the midst of shipping a major version update 😅

Jonathhhan commented:
> From the discussion above I think the general agreement was to focus on @juj 's Wasm Workers PR and to add AudioWorklet support on top of that? If so, then helping in that direction would probably be better. Or is there not consensus on that direction?

I agree totally with that (only that my programming knowledge is a little limited - but I am happy to test), and it would be great to have audioWorklet in the official branch. Its just, that at the moment this branch is the only I know where I can use audioWorklets together with Open Frameworks.

juj commented:
Wasm Workers PR did land, although not sure if that was premature ( https://github.com/emscripten-core/emscripten/pull/12833#issuecomment-1061645381  ). The PR #16449 is now up for reviewing and landing Audio Worklet support as well.

Looking at the dates on the comment history here, I see it has taken quite some time, so thanks for the patience on this one! Please give the above PR a review for any comments you might have.

yogesh0509 commented:
Hello everyone. I am new to open source and have arrived on this pr by following the guidelines mentioned on the Chromium GSOC 2022 Project Proposal: Audio Worklet for Emscripten. I would really like to contribute to this project. Can somebody brief me how I can contribute to this thread or project.

brainsucker committed:
fixed wasmTableMirror with ABORT_ON_WASM_EXCEPTIONS

tommitytom commented:
Hey folks, any update as to whats going on here?  WASM workers appear to have been merged - should they be used to handle this instead, or is this PR still the best method?

sbc100 commented:
Can this PR be closed now that #16449 has landed?

tklajnscek commented:
Agreed!
